data.md

```diff
@@ -273,9 +273,9 @@ The `#wrap` function wraps an integer to a given bit width.
  // -----------------------------------------
     rule #chop(< ITYPE > N) => < ITYPE > (N modInt #pow(ITYPE))
 
-    syntax Int  ::= #wrap(Int, Int) [function]
- // ------------------------------------------
-    rule #wrap(WIDTH, N) => N modInt (1 <<Int WIDTH) [concrete]
+    syntax Int  ::= #wrap(Int, Int) [function, functional]
+ // ------------------------------------------------------
+    rule #wrap(WIDTH, N) => N &Int ((1 <<Int WIDTH) -Int 1) [concrete]
```

Comment:
Pretty sure `#wrap` is not total because `<<Int` is not total (so you can't do the `functional` attribute).

But you can make it total by doing something like:

```
    rule #wrap(WIDTH, N) => N &Int ((1 <<Int WIDTH) -Int 1) requires         WIDTH >=Int 0 [concrete]
    rule #wrap(WIDTH, N) => N &Int ((1 <<Int 0    ) -Int 1) requires notBool WIDTH >=Int 0 [concrete]
```

Reply:
I'm thinking which makes most sense. I'd prefer make it non-total here, because it wouldn't really make sense for any step of the semantics to try to wrap with a negative number. But the prover might now play well with that.

@ttuegel Let's say we do the following:
```
syntax Int  ::= #wrap(Int, Int) [function]
 // --------------------------------------
    rule #wrap(WIDTH, N) => N &Int ((1 <<Int WIDTH) -Int 1) requires WIDTH >=Int 0 [concrete]
```
We then run the prover on a spec, where we end up with a `#wrap(WIDTH, N)` somewhere, and from the sideconditions in that proof branch we know that `WIDTH >=Int 0`. Would the prover know not to split on `#Ceil(#wrap(WIDTH, N))`? Or would it still consider it a possibility that `#wrap(WIDTH, N)` in undefined?

Reply:
Well, the way I see it, if you can make `#wrap` total and it still gives the correct semantic behavior on all concrete inputs, it just makes the prover able to handle the construct butter. The fact that it won't work in an intuitive way on negative numbers doesn't really matter, because that's not a use-case we ever need to prove for.

Reply:
Okay, that's reasonable. I'm still curious about how the prover would handle the case in my comment above, but making `#wrap` total is probably the better option.

I made some slight changes to your suggestion. For one thing, we may as well simplify the RHS when we wrap by non-positive numbers, because it will always just be 0. Second, I made that rule apply to the case where `WIDTH` is 0, too. Lastly, I made the second rule non-`concrete`, because I don't see why it should be when the entire RHS is concrete.